### ISSUE TITLE
Ubuntu 16.04 http_proxy not work

### ISSUE THREAD

**Issue by mhohai** [docker/docker#22568]
**Output of `docker version`:**

```
Docker version 1.11.1, build 5604cbe
```

**Output of `docker info`:**

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 6
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-22-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.734 GiB
Name: rooot
ID: 2EOA:RQJS:MIO3:BJA2:UFFW:QMRR:DOSS:6IVG:CAEB:N47O:2VVE:7XFH
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Steps to reproduce the issue:**
1. `vim /etc/default/docker` enable http_proxy
2. `docker pull ubuntu`
3. 

**Describe the results you received:**
I'm using the http_proxy powered by squid3 publish this issue, but pull image not using proxy.

**Describe the results you expected:**
http_proxy not work

**Additional information you deem important (e.g. issue happens only occasionally):**
socks proxy better the http proxy

**Comment by mhohai** [comment]
I'm install docker by deb: [deb https://apt.dockerproject.org/repo ubuntu-xenial main](https://docs.docker.com/engine/installation/linux/ubuntulinux/)

**Comment by thaJeztah** [comment]
The `/etc/default/docker` file is only used on systems using sysvinit and upstart (see this [comment at the top of the file](https://github.com/docker/docker/blob/master/contrib/init/sysvinit-debian/docker.default#L3-L8)). Ubuntu 16.04 uses systemd, so doesn't use that file.

You can use a `daemon.json` Configuration file, as described here;
https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file

Or a systemd drop-in file, following the instructions here;
https://docs.docker.com/engine/admin/systemd/#http-proxy

**Comment by thaJeztah** [comment]
Oh, actually, I don't think you can configure a proxy through the daemon.json, so ignore that

**Comment by mhohai** [comment]
thanks a lot

**Comment by thaJeztah** [comment]
I'll close this issue, but feel free to comment after I closed; hope this resolves your issue

**Comment by mhohai** [comment]
you close quickly than me. :smiley: 
but, can you tell me `is docker support socks proxy?`

**Comment by thaJeztah** [comment]
I think it was added in 1.11 through this PR; https://github.com/docker/docker/pull/20366 (but I haven't used it myself)

**Comment by strfx** [comment]
Just came across the same issue. However, I discovered that `/etc/default/docker` still works under Ubuntu 16.04 - even when using systemd. The systemd unit file under `/lib/systemd/system/docker.service` actually configures `/etc/default/docker` as `EnvironmentFile`, meaning it gets environment variables from this file.

Put your proxy settings into the file, but leave out the `export`, e.g.

```
# Docker Upstart and SysVinit configuration file
[...]
HTTP_PROXY="http://myawesomeproxy/"
HTTPS_PROXY="http://myawesomeproxy/"
```

Restart the Docker service `systemctl docker` and you should be fine.

**Comment by thaJeztah** [comment]
@strfx the official packages (apt.dockerproject.org) don't but it's possible the packages as shipped by Ubuntu add those lines (`EnvironmentFile`)

**Comment by strfx** [comment]
@thaJeztah Ah, you're right. Thanks for the info :)

**Comment by tonysongtl** [comment]
Where the username and password for the proxy in /etc/systemd/system/docker.service.d/http-proxy.conf?
[Service]
Environment="HTTP_PROXY=http://proxy.example.com:80/"

**Comment by thaJeztah** [comment]
@tonysongtl `HTTP_PROXY =http://USERNAME:PASSWORD@PROXYIP:PROXYPORT`

**Comment by tonysongtl** [comment]
Thanks.

I just configured as HTTPS_PROXY=http://USERNAME:PASSWORD@PROXYIP:PROXYPORT, but I got the error as follows:

Error while pulling image: Get https://index.docker.io/v1/repositories/library/hello-world/images: http: error connecting to proxy http://username:password@proxy.example.com:8080/: dial tcp: lookup username on 10.72.255.100:53: no such host

**Comment by mhohai** [comment]
@tonysongtl replace to yourself proxy for upper char: HTTP_PROXY=http://USERNAME:PASSWORD@PROXYIP:PROXYPORT

**Comment by tonysongtl** [comment]
@mhohai Thanks. 
I did replace the username, password and proxy address.  I deleted the real information here. It seems that Docker treat the username as the host, It doesn't support the format http://USERNAME:PASSWORD@PROXY:PORT

**Comment by tonysongtl** [comment]
The version of docker and OS:
docker --version
Docker version 1.12.4, build 1564f02
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04 LTS
Release:        16.04
Codename:       xenial

**Comment by tonysongtl** [comment]
@thaJeztah @mhohai 
Thank you,  I have resolved the problem.
The reason is that there is a special character in the password. I used %23 to replace the #,  it worked!

**Comment by thg303** [comment]
I have same problem but it is not solved.
Docker version 1.13.0, build 49bf474
from official docker repository.

I followed the instruction, I created:
/etc/systemd/system/docker.service.d/http-proxy.conf
also I tried changing **/etc/defaults/docker**
I can verify that Environment is set, but docker does not use the proxy:
systemctl show --property=Environment docker
Environment=HTTP_PROXY=http://127.0.0.1:7777/ http_proxy=http://127.0.0.1:7777/

I am using ubuntu open stack image:
http://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img

any Idea how may I track this bug or fix it?

**Comment by thaJeztah** [comment]
> I can verify that Environment is set, but docker does not use the proxy:

@thg303 does the proxy show up in `docker info`, e.g. does the output show something like this; ?

```
Name: debian-2gb-ams3-01
ID: 6SSD:M4SI:W3UM:7WCO:6WFG:67EM:KPVC:6Y4G:7GNI:ARNJ:KZSO:6I4X
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://127.0.0.1:7777
Registry: https://index.docker.io/v1/
```

**Comment by thg303** [comment]
@thaJeztah yes. sorry I found that "Registery" url was not in the list of my proxy. Now everything works fine.

**Comment by r1k0** [comment]
Hello,

I've encountered the same proxy issue on a freshly installed ubuntu xenial this weekend but I could only make it work the following way:

```
root@labint810:~# docker --version
Docker version 1.13.1, build 092cba3
root@labint810:~# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.2 LTS
Release:	16.04
Codename:	xenial
root@labint810:~# 
```

I've followed the official doc at https://docs.docker.com/engine/admin/systemd/#http-proxy but I was out of luck.
Adding this:
```
[Service]
Environment="HTTP_PROXY=http://proxy.example.com:80/"
```

to /etc/systemd/system/docker.service.d/http-proxy.conf and reloading systemd just didn't work.

I kept getting:
```
root@labint810:~# docker info | grep -i proxy
WARNING: No swap limit support
root@labint810:~# systemctl show --property=Environment docker
Environment=
root@labint810:~# cat /etc/systemd/system/docker.service.d/http-proxy.conf
[Service]
Environment=http_proxy=XXXXXX/
Environment=https_proxy=xxxxxxxx/
root@labint810:~#
```

So I went to /lib/systemd/system/docker.service and added:
```
EnvironmentFile=/etc/default/docker
```
to the [Service] block. 
I've added my proxy config to /etc/default/docker (without EnvironmentFile, config wouldn't be picked up) *without* exporting it

```
root@labint810:~# cat /etc/default/docker 
# Docker Upstart and SysVinit configuration file

#
# THIS FILE DOES NOT APPLY TO SYSTEMD
#
#   Please see the documentation for "systemd drop-ins":
#   https://docs.docker.com/engine/admin/systemd/
#

# Customize location of Docker binary (especially for development testing).
#DOCKERD="/usr/local/bin/dockerd"

# Use DOCKER_OPTS to modify the daemon startup options.
#DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4"

# If you need Docker to use an HTTP proxy, it can also be specified here.
#export http_proxy="http://127.0.0.1:3128/"

# This is also a handy place to tweak where Docker's temporary files go.
#export DOCKER_TMPDIR="/mnt/bigdrive/docker-tmp"

http_proxy="xxx"
https_proxy="xxx"
HTTP_PROXY="xxx"
HTTPS_PROXY="xxx"
root@labint810:~# 

```

I restarted docker and now I get the proper config:
```
root@labint810:~# docker info | grep -i proxy
WARNING: No swap limit support
Http Proxy: http://labordomainFGH546:Int3rn3t001@10.90.126.197:8080/
Https Proxy: http://labordomainFGH546:Int3rn3t001@10.90.126.197:8080/
root@labint810:~# 
```

I hope it helps someone as this kept me busy for a few hours.

**Comment by thaJeztah** [comment]
@r1k0 I am unable to reproduce your issue. I created a drop in file;

```bash
$ cat /etc/systemd/system/docker.service.d/http-proxy.conf
[Service]
Environment="HTTP_PROXY=http://proxy.example.com:80/"
Environment="HTTPS_PROXY=https://proxy.example.com:443/"
```

Did a `daemon reload` and restarted the docker service;

```bash
$ sudo systemctl daemon reload
$ sudo systemctl restart docker
```

Verified that the settings were picked up by `systemd`;

```bash
$ sudo systemctl show --property=Environment docker
Environment=HTTP_PROXY=http://proxy.example.com:80/ HTTPS_PROXY=https://proxy.example.com:443/
```

And in `docker info`;

```bash
$ sudo docker info | grep -i proxy
Http Proxy: http://proxy.example.com:80/
Https Proxy: https://proxy.example.com:443/
WARNING: No swap limit support
```

**Comment by r1k0** [comment]
@thaJeztah 

Thank you for your feedback.
I see my mistake.

Restarting docker is a requirement before checking the environment variable. Now that I think of it it makes sense and I should have thought about it! Shame on me. Since I could not validate step 5 I did not consider step 6.

My suggestion is to switch step 5 and 6 in the doc. Current step 6 (restart docker) should be current step 5 (show environment). Docker needs to be restarted before the Environment variable is read.

So the doc reads:

```
5) Restart Docker:

$ sudo systemctl restart docker

6) Verify that the configuration has been loaded:

$ systemctl show --property=Environment docker
Environment=HTTP_PROXY=http://proxy.example.com:80/
```

Thank again for your input, it helped me see what went wrong. I reverted my change and did it the proper way.
